x-pack/legacy/plugins/transform/public/shim.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage, FormattedDate, FormattedTime } from '@kbn/i18n/react';
+import { I18nContext } from 'ui/i18n';
+
+import chrome from 'ui/chrome';
+import { DOC_LINK_VERSION, ELASTIC_WEBSITE_URL } from 'ui/documentation_links';
+import { management, MANAGEMENT_BREADCRUMB } from 'ui/management';
+import { fatalError, toastNotifications } from 'ui/notify';
+import routes from 'ui/routes';
+import { docTitle } from 'ui/doc_title/doc_title';
+
+import { HashRouter } from 'react-router-dom';
+
+// @ts-ignore: allow traversal to fail on x-pack build
+import { createUiStatsReporter } from '../../../../../src/legacy/core_plugins/ui_metric/public';
+import { SavedSearchLoader } from '../../../../../src/legacy/core_plugins/kibana/public/discover/types';
+
+export interface AppCore {
+  chrome: typeof chrome;
+  i18n: {
+    [i18nPackage: string]: any;
+    Context: typeof I18nContext;
+    FormattedMessage: typeof FormattedMessage;
+    FormattedDate: typeof FormattedDate;
+    FormattedTime: typeof FormattedTime;
+  };
+  notification: {
+    fatalError: typeof fatalError;
+    toastNotifications: typeof toastNotifications;
+  };
+  savedSearches: {
+    getClient(): any;
+    setClient(client: any): void;
+  };
+}
+
+export interface AppPlugins {
+  management: {
+    sections: typeof management;
+  };
+}
+
+export interface Core extends AppCore {
+  chrome: typeof chrome;
+  http: {
+    getClient(): any;
+    setClient(client: any): void;
+  };
+  routing: {
+    registerAngularRoute(path: string, config: object): void;
+    registerRouter(router: HashRouter): void;
+    getRouter(): HashRouter | undefined;
+  };
+  documentation: {
+    esDocBasePath: string;
+    esPluginDocBasePath: string;
+    esStackOverviewDocBasePath: string;
+  };
+  docTitle: {
+    change: typeof docTitle.change;
+  };
+}
+
+export interface Plugins extends AppPlugins {
+  management: {
+    sections: typeof management;
+    constants: {
+      BREADCRUMB: typeof MANAGEMENT_BREADCRUMB;
+    };
+  };
+  uiMetric: {
+    createUiStatsReporter: typeof createUiStatsReporter;
+  };
+}
+
+export function createPublicShim(): { core: Core; plugins: Plugins } {
+  // This is an Angular service, which is why we use this provider pattern
+  // to access it within our React app.
+  let httpClient: ng.IHttpService;
+  // This is an Angular service, which is why we use this provider pattern
+  // to access it within our React app.
+  let savedSearches: SavedSearchLoader;
+
+  let reactRouter: HashRouter | undefined;
+
+  return {
+    core: {
```

Comment:
Many of these APIs are already available in the New Platform so you should be able to use them directly by importing from `ui/new_platform` rather than their legacy counterparts.

- `http` -> `npStart.core.http`
- `chrome` -> `npStart.core.chrome` (some APIs have been moved around, so you may still need to use the legacy service for some things (such as `chrome.getInjected`).
- `notification` -> `npStart.core.notifications` (does not support fatalError at this time).
- `documentation` -> `npStart.core.docLinks`

Reply:
You may want to make your shim an extension of `npStart.core`:

```js
return {
  core: {
    ..npStart.core,
     docTitle: { ... },
  }
}
```

Reply:
I addressed most of your suggestions in 8e0c5bd. Migrating to `npStart.core.http` I will do in a follow-up, I added a note to #45352.